File: day21.lua
#!/usr/bin/env lua

local X = -1
local R = 0
local D = 1
local L = 2
local U = 3
local A = 10

local dstr = { ">", "v", "<", "^" }

local function coord(o)
    return { x = o[1], y = o[2] }
end

local function state(o)
    local s = { pos = {}, num = 0 }
    if o then
        table.move(o.pos, 1, #o.pos, 1, s.pos)
        s.num = o.num
    end
    return s
end

local visit = {}

local function clear()
    visit = setmetatable({}, {
        __index = function(t, k)
            t[k] = {
                dist = math.maxinteger,
                nexts = state(),
                path = {},
            }
            return t[k]
        end
    })
end

local function lookup(s, level, d)
    local x = s.pos[level].x - 1
    local y = s.pos[level].y - 1
    local k = x | (y << 2) | (d << 3) | (level << 7)
    return visit[k]
end

local function dirvalid(x, y)
    return (y == 1 and x >= 2 and x <= 3)
        or (y == 2 and x >= 1 and x <= 3)
end

local function numvalid(x, y)
    return (y >= 1 and y <= 3 and x >= 1 and x <= 3)
        or (y == 4 and x >= 2 and x <= 3)
end

local dirpad = {
    { X, U, A },
    { L, D, R },
}

local invdirpad = {}
for y = 1, #dirpad do
    for x = 1, #dirpad[y] do
        local d = dirpad[y][x]
        if d ~= X then
            invdirpad[d] = coord{ x, y }
        end
    end
end

local numpad = {
    { 7, 8, 9 },
    { 4, 5, 6 },
    { 1, 2, 3 },
    { X, 0, A },
}

local invnumpad = {}
for y = 1, #numpad do
    for x = 1, #numpad[y] do
        local d = numpad[y][x]
        if d ~= X then
            invnumpad[d] = coord{ x, y }
        end
    end
end

local function sign(x)
    if x < 0 then return -1 end
    if x > 0 then return 1 end
    return 0
end

local nlevel
local code = {}

local solvelevel
local dirs = { { L, R }, { U, D } }

local function solvemove(s, level, tpos, firstaxis, path)
    local dist = 0
    local pos = s.pos[level]
    local posv = { pos.x, pos.y }
    local tposv = { tpos.x, tpos.y }
    local sdirv = { sign(tpos.x - pos.x), sign(tpos.y - pos.y) }
    for i = 1, 2 do
        local axis = firstaxis == i and 1 or 2
        while posv[axis] ~= tposv[axis] do
            posv[axis] = posv[axis] + sdirv[axis]
            if (level == nlevel and not numvalid(posv[1], posv[2]))
            or (level ~= nlevel and not dirvalid(posv[1], posv[2])) then
                return math.maxinteger
            end
            dist = dist + solvelevel(s, level - 1, sdirv[axis] < 0 and dirs[axis][1] or dirs[axis][2], path)
        end
    end
    s.pos[level] = tpos
    local adist = solvelevel(s, level - 1, A, path)
    dist = dist + adist
    return dist
end

solvelevel = function(s, level, d, path)
    if level < 1 then
        --[[
        if nlevel <= 3 then
            local buf = d == A and 'A' or dstr[d + 1]
            table.insert(path, buf)
        end
        --]]
        return 1
    end
    local v = nil
    if level < nlevel then
        v = lookup(s, level, d)
        if v.dist ~= math.maxinteger then
            table.move(v.nexts.pos, 1, level, 1, s.pos)
            table.move(v.path, 1, #v.path, #path + 1, path)
            return v.dist
        end
    end

    local tpos = (level == nlevel and invnumpad or invdirpad)[d]
    local nexts = { state(s), state(s) }
    local dists = { math.maxinteger, math.maxinteger };
    local paths = { {}, {} }
    for axis = 1, 2 do
        dists[axis] = solvemove(nexts[axis], level, tpos, axis, paths[axis])
    end

    assert(dists[1] < math.maxinteger or dists[2] < math.maxinteger)
    local min = dists[1] < dists[2] and 1 or 2
    table.move(nexts[min].pos, 1, level, 1, s.pos)
    table.move(paths[min], 1, #paths[min], #path + 1, path)

    if v then
        v.dist = dists[min]
        table.move(s.pos, 1, level, 1, v.nexts.pos)
        table.move(paths[min], 1, #paths[min], 1, v.path)
    end
    return dists[min]
end

local function solve(nbots)
    nlevel = nbots
    clear()

    local dist = 0
    local path = {}
    local s = state()
    for i = 1, nlevel - 1 do
        s.pos[i] = coord{ 3, 1 }
    end
    s.pos[nlevel] = coord{ 3, 4 }
    s.num = 1

    while s.num <= 4 do
        local target = code[s.num]
        dist = dist + solvelevel(s, nlevel, target, path)
        s.num = s.num + 1
    end
    if #path > 0 then print("path", table.concat(path)) end
    return dist
end

local input = io.open("day21.txt"):read("a")
local sum = { 0, 0 }

for xs in input:gmatch("(%d+)A") do
    local x = tonumber(xs)
    --print(string.format("%03dA", x))
    code[1] = (x // 100) % 10
    code[2] = (x //  10) % 10
    code[3] = (x //   1) % 10
    code[4] = A
    local d1 = solve(2 + 1)
    local d2 = solve(25 + 1)
    --print(d1, d2)
    sum[1] = sum[1] + d1 * x
    sum[2] = sum[2] + d2 * x
end

print(sum[1], sum[2])

File: day21.c
#include <assert.h>
#include <limits.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

#define COUNT(a) (sizeof(a) / sizeof((a)[0]))

enum {
    X = -1,
    R = 0,
    D = 1,
    L = 2,
    U = 3,
    A = 10,

    NBOTS = 25 + 1,
};

const char dstr[] = ">v<^";

typedef union {
    struct {
        int x, y;
    };
    int v[2];
} coord_t;

typedef struct {
    coord_t pos[NBOTS];
    int num;
} state_t;

#define KEYBITS (2 + 1 + 4 + 5)
typedef union {
    struct {
        unsigned x : 2;
        unsigned y : 1;
        unsigned d : 4;
        unsigned l : 5;
    };
    unsigned v;
} key_t;

typedef struct {
    long long dist;
    state_t nexts;
    char path[16];
} visit_t;

visit_t visit[1 << KEYBITS];

void clear()
{
    for (int i = 0; i < COUNT(visit); i++) {
        visit[i] = (visit_t){ .dist = LONG_LONG_MAX };
    }
}

visit_t* lookup(state_t* s, int level, int d)
{
    key_t k = {
        .x = s->pos[level].x,
        .y = s->pos[level].y,
        .d = d,
        .l = level,
    };
    return &visit[k.v];
}

bool dirvalid(int x, int y)
{
    return (y == 0 && x >= 1 && x <= 2)
        || (y == 1 && x >= 0 && x <= 2);
}

bool numvalid(int x, int y)
{
    return (y >= 0 && y <= 2 && x >= 0 && x <= 2)
        || (y == 3 && x >= 1 && x <= 2);
}

const int dirpad[2][3] = {
    { X, U, A },
    { L, D, R },
};

coord_t invdirpad[11];

const int numpad[4][3] = {
    { 7, 8, 9 },
    { 4, 5, 6 },
    { 1, 2, 3 },
    { X, 0, A },
};

coord_t invnumpad[11];

int sign(int x)
{
    if (x < 0) return -1;
    if (x > 0) return 1;
    return 0;
}

int nlevel;
int code[4];

long long solvelevel(state_t* s, int level, int d, char* path);

long long solvemove(state_t* s, int level, coord_t tpos, int firstaxis, char* path)
{
    const int dirs[2][2] = { { L, R }, { U, D } };
    long long dist = 0;
    coord_t* pos = &s->pos[level];
    coord_t sdir = {{ sign(tpos.x - pos->x), sign(tpos.y - pos->y) }};
    for (int i = 0; i < 2; i++) {
        int axis = i ^ firstaxis;
        while (pos->v[axis] != tpos.v[axis]) {
            pos->v[axis] += sdir.v[axis];
            if ((level == nlevel && !numvalid(pos->x, pos->y)) ||
                (level != nlevel && !dirvalid(pos->x, pos->y))) {

                return LONG_LONG_MAX;
            }
            dist += solvelevel(s, level - 1, sdir.v[axis] < 0 ? dirs[axis][0] : dirs[axis][1], path);
        }
    }
    dist += solvelevel(s, level - 1, A, path);;
    return dist;
}

long long solvelevel(state_t* s, int level, int d, char* path)
{
    if (level < 0) {
        #if 0
        if (nlevel < 3) {
            char buf[2] = { d == A ? 'A' : dstr[d], 0 };
            strcat(path, buf);
        }
        #endif
        return 1;
    }
    visit_t* v = NULL;
    if (level < nlevel) {
        v = lookup(s, level, d);
        if (v->dist != LONG_LONG_MAX) {
            memcpy(s->pos, v->nexts.pos, (level + 1) * sizeof(s->pos[0]));
            strcat(path, v->path);
            return v->dist;
        }
    }

    coord_t tpos = level == nlevel ? invnumpad[d] : invdirpad[d];
    state_t nexts[2] = { *s, *s };
    long long dists[2] = { LONG_LONG_MAX, LONG_LONG_MAX };
    char paths[2][64] = { "", "" };
    for (int axis = 0; axis < 2; axis++) {
        dists[axis] = solvemove(&nexts[axis], level, tpos, axis, paths[axis]);
    }

    assert(dists[0] < LONG_LONG_MAX || dists[1] < LONG_LONG_MAX);
    int min = dists[0] < dists[1] ? 0 : 1;
    *s = nexts[min];
    strcat(path, paths[min]);

    if (v) {
        v->dist = dists[min];
        memcpy(v->nexts.pos, s->pos, (level + 1) * sizeof(s->pos[0]));
        strcpy(v->path, paths[min]);
    }
    return dists[min];
}

long long solve(int nbots)
{
    nlevel = nbots - 1;
    clear();

    long long dist = 0;
    char path[128] = "";
    state_t s = { 0 };
    for (int i = 0; i < nlevel; i++) {
        s.pos[i] = (coord_t){{ 2, 0 }};
    }
    s.pos[nlevel] = (coord_t){{ 2, 3 }};
    s.num = 0;

    while (s.num < 4) {
        int target = code[s.num];
        dist += solvelevel(&s, nlevel, target, path);
        s.num++;
    }
    if (path[0]) printf("path %s\n", path);
    return dist;
}

int main()
{
    FILE* file = fopen("day21.txt", "r");
    long long sum = 0, sum2 = 0;

    for (int y = 0; y < COUNT(numpad); y++) {
        for (int x = 0; x < COUNT(numpad[y]); x++) {
            int d = numpad[y][x];
            if (d != X) {
                invnumpad[d] = (coord_t){{ x, y }};
            }
        }
    }

    for (int y = 0; y < COUNT(dirpad); y++) {
        for (int x = 0; x < COUNT(dirpad[y]); x++) {
            int d = dirpad[y][x];
            if (d != X) {
                invdirpad[d] = (coord_t){{ x, y }};
            }
        }
    }

    for (int x = 0; 1 == fscanf(file, "%dA", &x); ) {
        //printf("%03dA\n", x);
        code[0] = (x / 100) % 10;
        code[1] = (x /  10) % 10;
        code[2] = (x /   1) % 10;
        code[3] = A;
        long long d = solve(2 + 1);
        long long d2 = solve(25 + 1);
        //printf("%lld %lld\n", d, d2);
        sum += d * x;
        sum2 += d2 * x;
    }

    printf("%lld %lld\n", sum, sum2);
}
